sematic/api/endpoints/request_parameters.py

```diff
@@ -294,6 +296,37 @@ def _none_if_empty(name: str) -> Optional[str]:
 
 
 def jsonify_error(error: str, status: HTTPStatus):
+    is_warning = False
+
+    # Since jsonify_error is explicitly called to handle the
+    # error, there is no uncaught exception to be handled by
+    # the "catch all" exception handler & logger. This ensures
+    # that we generate a log message even for these explicitly
+    # handled errors, so we can triage more easily.
+    if status is None:
+        logger.error("Attempting to create an error without a code: %s", error)
+    elif 400 <= status.value < 500:
+        logger.warning("Bad request error: '%s'", error, stack_info=True)
+        is_warning = True
+    else:
+        logger.exception("Server error: '%s'", error)
+
+    # The above prints the error summary and stack to *this* call.
+    # We also want to show the stack trace for the root cause exception,
+    # if there indeed was one.
+    exception_type, exception, trace = sys.exc_info()
+    if exception is None:
+        logger.warning("No exception cause")
```

Comment:
Currently in the codebase:
- 500s are always called from `except` blocks. `logger.exception(...)` and `sys.exc_info()` will also retrieve the exception - and do that inefficiently, as the stack trace is retrieved a total of 3 times on this flow. Line 328 of this PR is unreachable.
- 400s can be called both from `except` blocks, and from custom validation logic; printing warnings that there was no exception cause is spammy, as this is the most common flow. Printing both the stack info and the exception info when the exception exists is redundant.

Stack lines must not be printed individually as they are part of the same logical message payload and splitting them up will result in other concurrent logs interweaving with these lines.

Please instead add a new parameter to this function to be able to pass the original exception. This parameter would default to None. This can be named `original_error`, or `cause`, etc. in order to make it clear that the error message can be custom, and this would be the actual cause. It can then be passed to the `exc_info` parameter in all log calls in this function.

The entire function would become:

```python
def jsonify_error(error: str, status: HTTPStatus, cause: Optional[BaseException] = None):
    if status is None:
        logger.error("Attempting to create an error without a code: %s", error)
    elif 400 <= status.value < 500:
        logger.warning("Bad request error: '%s'", error, exc_info=cause, stack_info=cause is None)
    else:
        logger.error("Server error: '%s'", error, exc_info=cause)

    return [...]
```